### ISSUE TITLE
Make gs://k8s-artifacts-prod regional

### ISSUE THREAD

**Issue by ameukam** [kubernetes/k8s.io#4303]
Currently the bucket is multi-regional:

```bash
❯ gcloud storage buckets describe gs://k8s-artifacts-prod --format 'table(name,location,locationType)'
NAME                LOCATION  LOCATION_TYPE
k8s-artifacts-prod  US        multi-region
```

with the new [pricing](https://cloud.google.com/storage/pricing-announce) for Cloud Storage applied in Oct 1, the bucket should be regional to minimize costs.

/priority important-soon
/area artifacts
/area release-eng
/milestone v1.26

**Comment by ameukam** [comment]
/assign @justinsb @hakman
the bucket is mainly used by kOps AFAIK.

**Comment by BenTheElder** [comment]
I don't think you can just make a bucket regional without delete and recreate?
Do we serve this behind an alias? If people are using the bucket directly we *cannot* move it non-disruptively.

https://cloud.google.com/storage/docs/moving-buckets

**Comment by ameukam** [comment]
> I don't think you can just make a bucket regional without delete and recreate? Do we serve this behind an alias? If people are using the bucket directly we _cannot_ move it non-disruptively.
> 
> [cloud.google.com/storage/docs/moving-buckets](https://cloud.google.com/storage/docs/moving-buckets)

The bucket is currently behind a GCLB: 
https://github.com/kubernetes/k8s.io/blob/ecee5b5712f9c18a154377354ab5efe09e443622/infra/gcp/bash/ensure-prod-storage-gclb.sh#L61-L66

The only possibility is to recreate to a newbucket as suggested in https://cloud.google.com/storage/docs/moving-buckets.

**Comment by hakman** [comment]
We are accessing the bucket via https://artifacts.k8s.io, in case it helps.

**Comment by BenTheElder** [comment]
maybe:
1. sync everything to a new regional bucket with a different name
2. Update artifacts.k8s.io to point to this
3. Delete and recreate the old bucket as regional (in case direct usage, somewhat disruptive)
4. switch artifacts.k8s.io back and dedupe buckets

**Comment by BenTheElder** [comment]
mostly finding hits for the GCR, but it's hard to rule out folks using the bucket directly.

**Comment by hakman** [comment]
Looking at the bucket contents, I only see it used by kOps and cri-tools. Though the latest version of cri-tools binaries is quite old:
```
% gsutil ls gs://k8s-artifacts-prod/binaries/             
gs://k8s-artifacts-prod/binaries/cri-tools/
gs://k8s-artifacts-prod/binaries/kops/

% gsutil ls gs://k8s-artifacts-prod/binaries/cri-tools
gs://k8s-artifacts-prod/binaries/cri-tools/v1.18.0/
gs://k8s-artifacts-prod/binaries/cri-tools/v1.19.0/
gs://k8s-artifacts-prod/binaries/cri-tools/v1.20.0/
gs://k8s-artifacts-prod/binaries/cri-tools/v1.21.0/
gs://k8s-artifacts-prod/binaries/cri-tools/v1.22.0/
```

**Comment by BenTheElder** [comment]
I think cri-tools binaries are canonically on github cc @saschagrunert

**Comment by saschagrunert** [comment]
> I think cri-tools binaries are canonically on github cc @saschagrunert

Yes, and we're additionally on `k8s-artifacts-cri-tools`

**Comment by ameukam** [comment]
Looking at:

```bash
❯ gcloud alpha storage buckets describe gs://k8s-artifacts-prod
etag: CJMB
iamConfiguration:
  bucketPolicyOnly:
    enabled: true
    lockedTime: '2019-10-07T22:09:04.125000+00:00'
  publicAccessPrevention: inherited
  uniformBucketLevelAccess:
    enabled: true
    lockedTime: '2019-10-07T22:09:04.125000+00:00'
id: k8s-artifacts-prod
kind: storage#bucket
location: US
locationType: multi-region
logging:
  logBucket: k8s-infra-artifacts-gcslogs
  logObjectPrefix: k8s-artifacts-prod
metageneration: '147'
name: k8s-artifacts-prod
projectNumber: '388270116193'
retentionPolicy:
  effectiveTime: '2019-08-16T17:20:11.468000+00:00'
  retentionPeriod: '315576000'
rpo: DEFAULT
selfLink: https://www.googleapis.com/storage/v1/b/k8s-artifacts-prod
storageClass: STANDARD
timeCreated: '2019-05-01T00:54:58.065000+00:00'
updated: '2022-08-10T17:22:10.414000+00:00'
website:
  mainPageSuffix: index.html
```

Logging is enabled
Uniform bucket-level access is enabled

**Comment by hakman** [comment]
Looks great. Thanks @ameukam! 😄

**Comment by ameukam** [comment]
About the IAM permissions:

```bash
 gsutil iam get gs://k8s-artifacts-prod | jq . | yq -y
bindings:
  - members:
      - group:k8s-infra-artifact-admins@kubernetes.io
      - projectEditor:k8s-artifacts-prod
      - projectOwner:k8s-artifacts-prod
    role: roles/storage.legacyBucketOwner
  - members:
      - projectViewer:k8s-artifacts-prod
    role: roles/storage.legacyBucketReader
  - members:
      - serviceAccount:k8s-infra-promoter@k8s-artifacts-prod.iam.gserviceaccount.com
    role: roles/storage.legacyBucketWriter
  - members:
      - group:k8s-infra-artifact-admins@kubernetes.io
      - serviceAccount:k8s-infra-promoter@k8s-artifacts-prod.iam.gserviceaccount.com
    role: roles/storage.objectAdmin
  - members:
      - allUsers
    role: roles/storage.objectViewer
etag: CJMB
```

The most important things to notice are:
k8s-infra-artifact-admins@kubernetes.io is empowered is:
- roles/storage.legacyBucketOwner
- roles/storage.objectAdmin

k8s-infra-promoter@k8s-artifacts-prod.iam.gserviceaccount.com is empowered with:
- roles/storage.objectAdmin
- roles/storage.legacyBucketWriter

**Comment by ameukam** [comment]
One approach discussed with @BenTheElder could be to:
- Create a new regional bucket `gs://k8s-infra-artifacts-prod` (or a better name. _Naming is hard_)
- actuate the IAM scripts with the new bucket
- ensure the GCLB (artifacts.k8s.io) use this bucket.
- update references of this bucket everywhere: https://cs.k8s.io/?q=gs%3A%2F%2Fk8s-artifacts-prod&i=nope&files=&excludeFiles=&repos=

**Comment by ameukam** [comment]
The Cloud gods listened to my prayers and we now have [buckets relocation](https://cloud.google.com/storage/docs/bucket-relocation/overview).

**Comment by saschagrunert** [comment]
@ameukam does it mean we can close this out? :)

**Comment by ameukam** [comment]
> [@ameukam](https://github.com/ameukam) does it mean we can close this out? :)

Once we do the migration. I want to do some testing to validate it's doable without breaking our users.
